### PR TITLE
Features/generic and specific solph components

### PR DESCRIPTION
**Pull request by ckaldemeyer**

### CODE REVIEW

**Comment by ckaldemeyer** [comment]
The current structure looks like the following:

* __init__.py
* blocks.py (responsible for blocks used in at least one component -> N:M relation)
* groupings.py (responsible for the grouping functionalities)
* inputlib/ (responsible for creating nodes -> currently contains the csv-reader)
  * __init__.py
  * csv_tools.py
* models.py  (responsible for the overall operational/invest/... model including sets, constraint and objective creation, methods to solve it and get additional information and result processing)
* network.py (responsible for inherited core and new solph classes and related functions)
* options.py  (responsible for classes that for some reason do not fit in network.py)
* plumbing.py (responsible for helper functions that should be provided for within other files)

An my sketched idea for a new structure looks like this:

* __init__.py
* main.py/model.py (responsible for abovementioned funcs + automatic component integration)
* components/ (responsible for components. network and block classes are stored within single modules)
  - [x] __init__.py
  * investment.py - in one "generic" module?
  * flow.py - in one "generic" module?
  * binaryflow.py - in one "generic" module?
  * discreteflow.py - in one "generic" module?
  * ...
  * sink.py - in one "generic" module? 
  * source.py - in one "generic" module?
  * transformer.py - in one "generic" module?
  * storage.py
  * some_fancy_component.py
* inputlib/ (responsible for creating nodes)
  * __init__.py
  * csv_tools.py
* misc/ (responsible for misc)
  * __init__.py
  * plumbing.py
  * groupings.py
* outputlib/ (responsible for creating results and integrating results for new components)
  * __init__.py
  * result_dict.py
  * result_dataframe.py

So the idea would be be that the only thing I have to do to add a new component (e.g. a CAES plant) is to create a new module within the package `components` and having everything "at a glance". Using this approach, I could also easily exchange components within single files, etc..

**Comment by ckaldemeyer** [comment]
I have mage up my mind again and did not find it easy to figure out the universal relation between the different classes that hold the attributes and the respective blocks. E. g. the generic network classes such as `Sink`and `Source` do not have blocks which breaks a clear 1:N or N:M relation and so on.

To me, there are either "generic" **network** classes such as `Sink`, `Source`, `1NTransformer` and `N1Transformer`, **special** components, **options** such as `BinaryFlow`, `Flow` etc. that are used within the aforedmentioned classes and **blocks** that provide models for one of the aforementioned.

So generally, the following package/module structure might be an idea:

- [ ] models.py (responsible for model creation, sets, ... and automatic block and network/option class integration)
- [ ] groupings.py (responsible for groupings)
- [ ] plumbing.py (responsible for stuff without a place. holds only sequence class -> as single module into options package?)
- [ ] blocks/ (responsible for solph building blocks)
  - __init__.py
  [- sink.py -> repr through flow without single model]
  [- source.py -> repr through flow without single model]
  - binaryflow.py
  - bus.py
  - discreteflow.py
  - flow.py
  - 1ntransformer.py
  - n1transformer.py
  - storage.py
  - some_fancy_component.py
- [ ] network/ (responsible for network classes)
  - __init__.py
  - sink.py
  - source.py
  - 1ntransformer.py
  - n1transformer.py
- [ ] special_components/ (responsible for special component classes)
  - __init__.py
  - storage.py
  - investmentstorage.py
  - some_fancy_component.py
- [ ] options/ (responsible for optional classes that are used in combination with network or special component (sub)classes)
  - __init__.py
  - binaryflow.py
  - discreteflow.py
  - flow.py
  - investment.py
  - investmentflow.py

- [x] inputlib/ (responsible for creating nodes)
  - __init__.py
  - csv_tools.py
- [x] ../outputlib/ (responsible for creating results and integrating results for new components)
  - __init__.py
  - result_dict.py
  - result_dataframe.py

One difference to my last proposal is that I have to add two files/modules for a new component (one block and one network/special component) and these two files are automatically integrated in `models.py` where the defined constraints are set as it is made now:

        # loop over all constraint groups to add constraints to the model
        for group in self._constraint_groups:
            # create instance for block
            block = group()
            # Add block to model
            self.add_component(str(block), block)

Based on @simnh s and my talk in DK, this kind of divides into generic and specific components but a have to look at two places if I want to add something.

@oemof/oemof-solph : What do you think about this?

If you like it, I'll start with the refactoring tomorrow..

**Comment by ckaldemeyer** [comment]
I don't know if `options` could also be renamed to `other` or `misc` since not everything is an option.

But generally I think splitting of the different classes into packages and modules with single categories/responsibilities is cleaner and more flexible for new components which can be integrated by creating two files. This allows for more components, reduces potential merge conflicts, ...

**Comment by uvchik** [comment]
1. I like the idea to make it easier to add new components but I do not understand why it is necessary to change the structure. Couldn't we just add a new module for higher components, wait how it works and then take time to rethink the structure. I think changing the structure take time and discussion. As I understand we do not have the time, so why do we need the restructure?

2. For me the flow is a very basic object and you cannot build an energy system without a flow therefore I would not search it in options/misc/other.

3. I do not understand the difference between `special_components` and `network`. From my point of view we create a network using components and flows. At the moments all components can be found in network.py and there blocks in blocks.py. I think it is clearer than the new structure.

4. `Investment` is in option-class that cannot be used alone but inside a flow or a storage. I would put classes option-classes together in one module. I know the flow class is very special, too but it is the heart of solph (see. 2).

5. The really confusing thing at the moment is that a storage calls the storage block. If the storage has an investment object it calls the investment storage block. If a flow has an investment object the flow block is called and the investment-flow-block on top not instead. What happens if I use the DiscreteFlow or BinaryFlow? Is an additional block called or an binary-flow-block instead? This is very difficult to understand for new users. At the beginning we decided, that it should be possible to add such optional class to every component or that we have to add an extra warning (something like: this component does not work with the investment object).

**Comment by uvchik** [comment]
What about a new module where one can add new components together with its block and leave the structure untouched.

I am not against a restructure but we should be sure that we do not replace one problem by another.

**Comment by ckaldemeyer** [comment]
> I like the idea to make it easier to add new components but I do not understand why it is necessary to change the structure.
Because now there are blown up files with heaps of classes that differ in their structure and kind.

> Couldn't we just add a new module for higher components, wait how it works and then take time to rethink the structure. I think changing the structure take time and discussion. As I understand we do not have the time, so why do we need the restructure?

I offered my time and its more realigning things than writing new stuff. But I would either start it now or create my own solution because I need some special components and I would prefer to do it in a new structure instead of going in the app-direction of HESYSOPT or working on an old and very customized branch.

> For me the flow is a very basic object and you cannot build an energy system without a flow therefore I would not search it in options/misc/other.

It depends on how you look at it. From the functionality side network components are connected via flows. Therefore for me it would be o.k. like this. But you can make a suggestion.

> I do not understand the difference between special_components and network. From my point of view we create a network using components and flows. At the moments all components can be found in network.py and there blocks in blocks.py. I think it is clearer than the new structure.

To me, there are some basic network components (sink, source, 1N, N1) and the rest (storage, fancyxy) is already special as there's more stuff. But I could also live with your "all components view". By the way, this was my first proposal!

> Investment is in option-class that cannot be used alone but inside a flow or a storage. I would put classes option-classes together in one module. I know the flow class is very special, too but it is the heart of solph (see. 2).

I would be fine with this as I said the name/place can be different.

> The really confusing thing at the moment is that a storage calls the storage block. If the storage has an investment object it calls the investment storage block. If a flow has an investment object the flow block is called and the investment-flow-block on top not instead. What happens if I use the DiscreteFlow or BinaryFlow? Is an additional block called or an binary-flow-block instead? This is very difficult to understand for new users.

Yes, and I am trying to get it a bit more straight.

> At the beginning we decided, that it should be possible to add such optional class to every component or that we have to add an extra warning (something like: this component does not work with the investment object).

O.k.

To summarize my point of view:

If we do not decide this thing until tomorrow I will just create my own app and we then somehow see how this can be changed at a later stage. There's quite some project stuff I have to do and I have to make progress in this. This was just an offer to tighten the structure a bit but it's also fine if you're happy like this.

Just let me know because then I can align my plans!

**Comment by uvchik** [comment]
For me the main question is how to deal with the options. I like the way with the options because sometime the options only add some constraints. That means that every component has one solph class but may have more than one block depending on the options. 

With this proposal every component has to note in its docstring which options are possible.

Flow: Investment, Discrete, Binary (no combination possible)
Storage: Investment
MySpecialComponent: Investment, Discrete
MyNewComponent: NewOption

That means that my_special_component may have four blocks: 
* normal
* with Investment
* with Discrete
* with Discrete and Investment

We may use tuples if special combinations are possible.

Flow: Investment, Discrete, Binary, (Investment, Discrete), (Binary, Discrete)

- [ ] models.py (responsible for model creation, sets, ... and automatic block and network/option class integration)
- [ ] groupings.py (responsible for groupings)
- [ ] plumbing.py (responsible for stuff without a place. holds only sequence class -> as single module into options package?)
- [ ] network/ (responsible for component classes and its blocks)
  - __init__.py
  - bus.py (solph-class + block)
  - sink.py (solph-class)
  - source.py (solph-class)
  - 1ntransformer.py (solph-class + block)
  - n1transformer.py (solph-class + block)
  - storage.py (solph-class + two blocks)
  - some_fancy_component.py (solph-class + x blocks)

- [ ] flows/ (responsible for flows and its blocks)
  - __init__.py
  - binaryflow.py (block)
  - discreteflow.py (block)
  - flow.py (solph-class + block)
  - investmentflow.py (block)

- [ ] options/ (if these options are present different blocks will be used)
  - __init__.py
  - investment.py (option-class)
  - binary (option-class)
  - discrete (option-class)

- [x] inputlib/ (responsible for creating nodes)
  - __init__.py
  - csv_tools.py

- [x] ../outputlib/ (responsible for creating results and integrating results for new components)
  - __init__.py
  - result_dict.py
  - result_dataframe.py

*******************

Another way is to remove the options and name every component by its options. That may lead to more duplicates because you have to repeat the same constraints again and again. If you have an error in any constraint you have to fix it in different blocks. On the other hand every component has one block.

* InvestmentExampleComponent (solph-class + one block)
* InvestmentDiscretExampleComponent (solph-class + one block)
* BinaryExampleComponente (solph-class + one block)

********************

I like the first variant more and it is closer to the actual structure (-> less work).

**Comment by ckaldemeyer** [comment]
> For me the main question is how to deal with the options. I like the way with the options because sometime the options only add some constraints. That means that every component has one solph class but may have more than one block depending on the options.
> 
> With this proposal every component has to note in its docstring which options are possible.
> 
> Flow: Investment, Discrete, Binary (no combination possible)
> Storage: Investment
> MySpecialComponent: Investment, Discrete
> MyNewComponent: NewOption
> 
> That means that my_special_component may have four blocks:
> (the blocks may be additional or exclusive)
> 
> normal
> with Investment
> with Discrete
> with Discrete and Investment
> We may use tuples if special combinations are possible.
> 
> Flow: Investment, Discrete, Binary, (Investment, Discrete), (Binary, Discrete)
> 
>  models.py (responsible for model creation, sets, ... and automatic block and network/option class integration)
> 
>  groupings.py (responsible for groupings)
> 
>  plumbing.py (responsible for stuff without a place. holds only sequence class -> as single module into options package?)
> 
>  network/ (responsible for component classes and its blocks)
> 
> init.py
> bus.py (solph-class + block)
> sink.py (solph-class)
> source.py (solph-class)
> 1ntransformer.py (solph-class + block)
> n1transformer.py (solph-class + block)
> storage.py (solph-class + two blocks)
> some_fancy_component.py (solph-class + x blocks)
>  flows/ (responsible for flows and its blocks)
> 
> init.py
> binaryflow.py (block)
> discreteflow.py (block)
> flow.py (solph-class + block)
> investmentflow.py (block)
>  options/ (if these options are present different blocks will be used)
> 
> init.py
> investment.py (option-class)
> binary (option-class)
> discrete (option-class)
>  inputlib/ (responsible for creating nodes)
> 
> init.py
> csv_tools.py
>  ../outputlib/ (responsible for creating results and integrating results for new components)
> 
> init.py
> result_dict.py
> result_dataframe.py
> Another way is to remove the options and name every component by its options. That may lead to more duplicates because you have to repeat the same constraints again and again. If you have an error in any constraint you have to fix it in different blocks. On the other hand every component has one block.
> 
> InvestmentExampleComponent (solph-class + one block)
> InvestmentDiscretExampleComponent (solph-class + one block)
> BinaryExampleComponente (solph-class + one block)
> I like the first variant more and it is closer to the actual structure (-> less work).

Thanks a lot for your feedback. I think sometimes it's just better to talk directly like we did yesterday. Even if all the discussion is sometimes annoying for one of both sides... :smile: 

I also like the first variant more and will start as soon as I get my hands on it. But today I will first get my hands on the wind and water.. :smile:

**Comment by uvchik** [comment]
> But today I will first get my hands on the wind and water.. :smile:

windpower and hydro - great :smirk:

**Comment by ckaldemeyer** [comment]
I have now created the packages `options`, `flows`and `network` but still left all blocks in `blocks.py` since I'd like to discuss how to proceed from here on. I'll think about it and make an approach tomorrow morning.. ;-)

**Comment by ckaldemeyer** [comment]
The problem is that so far all blocks are named identically to their network/option/flow class and are generated like this in `models.py`:

        # ########################### CONSTRAINTS #############################
        # loop over all constraint groups to add constraints to the model
        for group in self._constraint_groups:
            # create instance for block
            block = group()
            # Add block to model
            self.add_component(str(block), block)
            # create constraints etc. related with block for all nodes
            # in the group
            block._create(group=self.es.groups.get(group))

There are different ways to deal with this now from renaming the block from `MyClass` to `MyClassBlock` (which does not look nice in the LP file) and other ways. As I said, I'll make a proposal..

**Comment by simnh** [comment]
I had some more thoughts on that an I think we you definitlely use this new strucutre. 

What I would add: 

* [ ] Create BaseModel in main.py
* [ ] Add specific models: Distpach, Investment, UnitCommitment, ... to models.py
* [ ] Inside add specfic components in  investmen.py etc that can be used with the models in models.py

I will add that here...

This is also related to #308 as this will be the general BaseModel would require an additional index. 

I will work on this on this branch!

**Comment by simnh** [comment]
Close this in favor of #329
